### ISSUE TITLE
return_linter() for implicit returns has a false positive on nested early returns

### ISSUE THREAD

**Issue by MichaelChirico** [r-lib/lintr#2354]
Corrected this test, and added the incorrect test as a new regression test.

That said, the new fix is more expedient than perfect -- it'll throw an incorrect lint here:

```r
foo <- function() {
  if (TRUE) {
    if (TRUE) {
      return(1)
    }
    2
  } else {
    3
  }
```

(the `return()` is a valid early return within that branch).

But as this issue is not relevant to the implicit `else` returns feature, I'm pushing to follow-up.

_Originally posted by @MichaelChirico in https://github.com/r-lib/lintr/pull/2321#discussion_r1405571314_

**Comment by MichaelChirico** [comment]
Test case for regression:

```r
test_that("explicit returns in control flow are linted", {
  lint_msg <- rex::rex("Use implicit return behavior")

  expect_lint(
    trim_some("
      foo <- function(bar) {
        if (TRUE) {
          return(bar)
        } else {
          return(NULL)
        }
      }
    "),
    list(lint_msg, lint_msg),
    return_linter()
  )
})
```

**Comment by MichaelChirico** [comment]
Also ensure that e.g. R/any_duplicated_linter.R does not lint.
